---
title: React appelle les composants et les Hooks
---

<Intro>

React s'occupe de rÃ©aliser le rendu des composants et des Hooks lorsque c'est nÃ©cessaire, afin d'optimiser l'expÃ©rience utilisateur. C'est une approche dÃ©clarativeÂ : vous dites Ã  React quel rendu effectuer dans la logique de votre composant, et React dÃ©terminera comment rÃ©aliser au mieux l'affichage pour l'utilisateur.

</Intro>

<InlineToc />

---

## N'appelez jamais les fonctions composants directement {/*never-call-component-functions-directly*/}

Les composants ne devraient Ãªtre utilisÃ©s que dans du JSX. Ne les appelez pas en tant que fonctions classiques. C'est Ã  React de les appeler.

C'est Ã  React de dÃ©cider quand appeler votre fonction composant [lors du rendu](/reference/rules/components-and-hooks-must-be-pure#how-does-react-run-your-code). En React, vous utilisez pour Ã§a la syntaxe JSX.

```js {2}
function BlogPost() {
  return <Layout><Article /></Layout>; // âœ… CorrectÂ : composants utilisÃ©s seulement par JSX
}
```

```js {2}
function BlogPost() {
  return <Layout>{Article()}</Layout>; // ğŸ”´ IncorrectÂ : ne les appelez jamais directement
}
```

Si un composant contient des Hooks, il est aisÃ© d'enfreindre les [RÃ¨gles des Hooks](/reference/rules/rules-of-hooks) lorsque les composants sont appelÃ©s directement au sein de boucles ou de conditions.

Le fait de laisser React orchestrer le rendu offre de nombreux avantagesÂ :

* **Les composants deviennent plus que de simples fonctions.** React peut leur ajouter des fonctionnalitÃ©s telles que _l'Ã©tat local_ au moyen de Hooks associÃ©s Ã  l'identitÃ© du composant dans l'arbre.
* **Les types des composants participent Ã  la rÃ©conciliation.** En laissant React appeler vos composants, vous lui en dites plus sur la structure sÃ©mantique de votre arbre. Par exemple, lorsque vous passez du rendu d'un `<Feed>` Ã  celui d'une page `<Profile>`, React ne risque pas de mÃ©langer leurs donnÃ©es sous-jacentes.
* **React peut amÃ©liorer l'expÃ©rience utilisateur.** Il peut par exemple laisser le navigateur traiter certaines tÃ¢ches entre les appels des composants, afin que le rendu d'un Ã©norme arbre de composants ne bloque pas le *thread* principal.
* **Le dÃ©bogage est facilitÃ©.** Si les composants sont des citoyens de plein droit connus de la bibliothÃ¨que, nous pouvons construire un outillage dÃ©veloppeurÂ·ses avancÃ© pour les examiner lors du dÃ©veloppement.
* **La rÃ©conciliation est plus efficace.** React peut dÃ©terminer prÃ©cisÃ©ment quels composants dans l'arbre ont besoin d'un nouveau rendu, ainsi que ceux qui peuvent Ãªtre laissÃ©s tels quels. Ã‡a aide Ã  rendre votre appli plus rapide et rÃ©active.

---

## Ne passez jamais des Hooks comme des valeurs classiques {/*never-pass-around-hooks-as-regular-values*/}

Les Hooks ne devraient Ãªtre appelÃ©s qu'au sein de composants. Ã‰vitez de les manipuler ou transmettre comme des valeurs classiques.

Les Hooks vous permettent d'ajouter des fonctionnalitÃ©s React Ã  un composant.  Ils devraient toujours Ãªtre appelÃ©s en tant que fonctions, jamais traitÃ©s comme des valeurs Ã  transmettre. Ã‡a permet un _raisonnement local_, c'est-Ã -dire que Ã§a offre la possibilitÃ© pour les dÃ©veloppeurÂ·ses de comprendre l'entiÃ¨retÃ© d'un composant grÃ¢ce Ã  son seul code source.

Enfreindre cette rÃ¨gle empÃªchera React d'optimiser automatiquement votre composant.

### Ne modifiez pas dynamiquement un Hook {/*dont-dynamically-mutate-a-hook*/}

Les Hooks devraient Ãªtre aussi Â«Â statiquesÂ Â» que possible.  Ã‡a signifie que vous ne devriez pas les modifier dynamiquement.  Vous devriez par exemple Ã©viter d'Ã©crire des Hooks d'ordre supÃ©rieurÂ :

```js {2}
function ChatInput() {
  const useDataWithLogging = withLogging(useData); // ğŸ”´ DÃ©conseillÃ©Â : Ã©vitez les Hooks dâ€™ordre supÃ©rieur
  const data = useDataWithLogging();
}
```

Les Hooks devraient Ãªtre immuables, et donc non modifiÃ©s.  PlutÃ´t que de modifier dynamiquement un Hook, crÃ©ez une version statique du Hook avec la fonctionnalitÃ© voulue.

```js {2,6}
function ChatInput() {
  const data = useDataWithLogging(); // âœ… CorrectÂ : utilise une nouvelle version du Hook
}

function useDataWithLogging() {
  // ... CrÃ©e une nouvelle version du Hook en intÃ©grant directement la logique ici
}
```

### N'utilisez pas dynamiquement les Hooks {/*dont-dynamically-use-hooks*/}

Les Hooks ne devraient par ailluers pas Ãªtre utilisÃ©s dynamique.  Disons par exemple que vous recouriez Ã  l'injection de dÃ©pendance en passant un Hook Ã  un composant sous forme de valeurÂ :

```js {2}
function ChatInput() {
  return <Button useData={useDataWithLogging} /> // ğŸ”´ IncorrectÂ : ne passez pas les Hooks en props
}
```

Vous devriez toujours intÃ©grer directement l'appel au Hook dans votre composant, et y traiter la logique associÃ©e.

```js {6}
function ChatInput() {
  return <Button />
}

function Button() {
  const data = useDataWithLogging(); // âœ… CorrectÂ : utilise le Hook directement
}

function useDataWithLogging() {
  // Si une logique conditionnelle modifie le comportement du Hook, elle devrait Ãªtre intÃ©grÃ©e
  // directement dans le Hook.
}
```

De cette faÃ§on, il est bien plus facile de comprendre et dÃ©boguer `<Button />`.  Lorsque les Hooks sont utilisÃ©s de faÃ§on dynamique, Ã§a augmente significativement la complexitÃ© de votre appli et empÃªche le raisonnement local, ce qui rÃ©duit sur le long terme la productivitÃ© de votre Ã©quipe.  Ã‡a augmente aussi le risque d'enfreindre par inadvertance les [RÃ¨gles des Hooks](/reference/rules/rules-of-hooks), qui dictent qu'un Hook ne doit jamais Ãªtre appelÃ© conditionnellement. Si vous vous retrouvez Ã  devoir *mocker* des composants pour vos tests, il vaut mieux *mocker* le serveur pour qu'il rÃ©ponde avec des donnÃ©es prÃ©dÃ©finies.  Lorsque c'est possible, il est gÃ©nÃ©ralement plus efficace de tester votre appli au moyen de tests *end-to-end* (E2E).
